lib/sync/syncer.go

```diff
@@ -91,52 +97,113 @@ func (s *Syncer) Stop() error {
 }
 
 func (s *Syncer) Start() error {
-	s.logger.Info("Starting syncer")
+	s.logger.Info("starting syncer")
 	s.workPool = NewPool(s.poolSize)
 	s.loop()
 	return nil
 }
 
+func (s *Syncer) SetSyncTargetBlock(ctx context.Context, height uint64) error {
+	select {
+	case s.updateHighestBlock <- height:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return nil
+}
+
+func (s *Syncer) SyncProgress(ctx context.Context) (*SyncProgress, error) {
+	c := make(chan *SyncProgress, 1)
+	select {
+	case s.getSyncProgress <- c:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	select {
+	case sp := <-c:
+		return sp, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
 func (s *Syncer) loop() {
 	checkc := s.afterFunc(s.checkInterval)
-	blockHeight := uint64(1)
+
+	height := s.lastestBlockHeight()
+	syncProgress := &SyncProgress{
+		StartingBlock: height,
+		CurrentBlock:  height,
+		HighestBlock:  height,
+	}
+
+	s.logger.Info("starting block to sync", "height", height)
+
 	for {
 		select {
 		case <-checkc:
-			blockHeight = s.syncBlockHeight(blockHeight)
+			syncProgress.HighestBlock++ // TODO(anarcher): Until work together consensus
+			s.sync(syncProgress)
 			checkc = s.afterFunc(s.checkInterval)
+		case height := <-s.updateHighestBlock:
+			if height >= syncProgress.CurrentBlock {
+				syncProgress.HighestBlock = height
+				s.sync(syncProgress)
+			}
+		case c := <-s.getSyncProgress:
+			c <- syncProgress.Clone()
 		case c := <-s.stop:
 			close(c)
 			return
 		}
 	}
 }
 
-func (s *Syncer) syncBlockHeight(height uint64) uint64 {
-	blk, err := block.GetLatestBlock(s.storage)
-	if err != nil {
-		s.logger.Error("checkBlockHeight", "err", err)
+func (s *Syncer) sync(p *SyncProgress) {
+	var (
+		startHeight        = p.CurrentBlock + 1
+		currentHeight      = p.CurrentBlock
+		highestHeight      = p.HighestBlock
+		lastestBlockHeight = s.lastestBlockHeight()
+		log                = func() {
+			s.logger.Info("sync progress",
+				"start", p.StartingBlock, "cur", p.CurrentBlock, "high", p.HighestBlock)
+		}
+	)
+
+	if lastestBlockHeight > p.CurrentBlock {
+		currentHeight = lastestBlockHeight
 	}
-	newHeight := blk.Height + 1
-	if newHeight > height {
-		s.work(newHeight)
-		s.logger.Info("Starting sync", "height", newHeight)
-		return newHeight
+	if currentHeight >= highestHeight {
+		log()
+		return
+	}
+
+	for height := startHeight; height <= highestHeight; height++ {
+		// TryAdd for unblocking when the pool is full. Just keep syncprogress for next sync
+		if s.work(height) == false {
+			break
+		}
+		currentHeight = height
 	}
-	return height
+	p.StartingBlock = startHeight
+	p.CurrentBlock = currentHeight
+	log()
 }
 
-func (s *Syncer) work(height uint64) {
+func (s *Syncer) work(height uint64) bool {
 	ctx := s.ctx
 	work := func() {
-		currHeight := s.currBlockHeight()
-		if currHeight > 0 && height <= currHeight {
-			s.logger.Info("This height has already synced", "height", height)
+		lastHeight := s.lastestBlockHeight()
```

Comment:
How about `latestHeight := s.latestBlockHeight()`?
`latestBlock` is used in consensus :)